K8s Network Model Diagrams

chrismetz09 opened the pull request:
Original issue described in issue #32243. Existing text was incomplete and required new text and diagrams to fully describe what the K8s network model enables using implementation examples.

mengjiao-liu commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,185 @@ description: >
+
+* IPv4 address - 32 bit IP addresses assigned to pod interfaces, host interfaces, L2 bridges, tunnel end-points and any other network resource requiring IPv4 connectivity. An example is `192.68.13.1`.
+  
+* IPv6 address - 128 bit IP address assigned to pod interfaces, host interfaces, L2 bridges, L3 routers, tunnel end-points and any other network resources requiring IPv6 connectivity. You can configure one or more IPv6 addresses on an interface. An example is `2001:db8:3333:4444:5555:6666:7777:8888.`
```
```suggestion
* IPv6 address - 128 bit IP address assigned to pod interfaces, host interfaces, L2 bridges, L3 routers, tunnel end-points and any other network resources requiring IPv6 connectivity. You can configure one or more IPv6 addresses on an interface. An example is `2001:db8:3333:4444:5555:6666:7777:8888`.
```

mengjiao-liu commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,185 @@ description: >
+  
+* L2bridge is a virtual L2 bridge that allows attachment, configuration and communications between pods on the same node. Depending on your deployment of container runtimes and CNI plugins, this bridge function could be implemented within the CNI plugin itself.  
+
+* [Network plugins](/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/#:~:text=A%20CNI%20plugin%20is%20required,releases%20of%20the%20CNI%20specification.) allow pods to communicate even when the source pod destination pod are running on different nodes. Network plugins achieve this in different ways. Networks plugins are also known as _CNI plugins_.
```
```suggestion
* [Network plugins](/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/) allow pods to communicate even when the source pod destination pod are running on different nodes. Network plugins achieve this in different ways. Networks plugins are also known as _CNI plugins_.
```
Is it necessary to point to something specific? What if the content of the article changes? I think it's enough to point directly to the article.

mengjiao-liu commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,185 @@ description: >
+
+Your containers within a pod must coordinate port usage, but this is no different from processes in a VM.  
+
+How you implement this is a detail of the particular container runtime you employ in your environment .
```
```suggestion
How you implement this is a detail of the particular container runtime you employ in your environment.
```

mengjiao-liu reviewed (commented):
Great job！ 
just some nits:

mengjiao-liu commented:
Technical details need to be confirmed by sig network reviewers.
/sig network

chrismetz09 commented:
/do-not-merge

chrismetz09 commented:
Tx for the review so far @mengjiao-liu. Adding more commits today and look forward to an updated review.

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+{{< figure src="/docs/images/k8net-net-arch2.drawio.svg" alt="k8s net arch" class="diagram-large" caption="Figure 1. K8s Network Architecture and Components" >}}
+The architecture and components consist of the following:
+
+* Nodes configured in virtual (VM) or physical (bare-metal) environments.
```
```suggestion
* Nodes, running as Linux virtual machines (Kubernetes also supports running on physical servers,
  and you can use either Windows or Linux as a node operating system)
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+
+* Nodes configured in virtual (VM) or physical (bare-metal) environments.
+  
+* Pods configured on each node with one or more containers.
```
```suggestion
* Pods running on each node with one or more containers. This example only has two pods in the cluster,
  to make the diagrams simpler.
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+  * You can assign an Ethernet mac address to each pod.
+    Kubernetes does not require that a Pod has a unique identity at the data-link layer.
+ 
+* Pods use a virtual point-to-point veth "link" between the pod network namespace and root network namespace. This allows pod packets to utilize network functions defined in the root network namespace.  
```
```suggestion
* The network plugin configures a virtual point-to-point veth link between the pod network namespace and
  the node's root network namespace. This allows pod packets to utilize network functions defined in the
  root network namespace (for example: health check traffic from the kubelet to the pod).
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+ 
+* Pods use a virtual point-to-point veth "link" between the pod network namespace and root network namespace. This allows pod packets to utilize network functions defined in the root network namespace.  
+  
+* L2bridge is a virtual L2 bridge that allows attachment, configuration and communications between pods on the same node. Depending on your deployment of container runtimes and CNI plugins, the bridge function could be implemented within the CNI plugin itself.  
```
```suggestion
* L2bridge is a virtual [layer 2](https://en.wikipedia.org/wiki/OSI_model#Layer_2:_Data_link_layer) bridge
  that allows attachment, configuration and communications between pods on the same node.
  Depending on your deployment of container runtimes and network plugins, the local packet forwarding
  function could be implemented within the network plugin itself.
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+  
+* L2bridge is a virtual L2 bridge that allows attachment, configuration and communications between pods on the same node. Depending on your deployment of container runtimes and CNI plugins, the bridge function could be implemented within the CNI plugin itself.  
+
+* [Network plugins](/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/) allow pods to communicate even when the source pod destination pod are running on different nodes. Network plugins achieve this in different ways. Networks plugins are also known as _CNI plugins_.
```
```suggestion
* A [_network plugin_](/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/) sets up
  container networking and allows pods to communicate even when the source pod destination pod are
  running on different nodes.
  Network plugins achieve this in different ways. Networks plugins are also known as _CNI plugins_.
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+
+* [Network plugins](/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/) allow pods to communicate even when the source pod destination pod are running on different nodes. Network plugins achieve this in different ways. Networks plugins are also known as _CNI plugins_.
+  
+* Virtual overlay network consists of a network of tunnels connecting pods running on different nodes.
```
```suggestion
* The virtual _overlay network_ consists of a network of tunnels connecting different nodes, using
   IP addresses from the IP address range that is allocated for use by pods. In this example, the
   node's host network adapter is in one address range, and pods are in a different address range.
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+  
+* Virtual overlay network consists of a network of tunnels connecting pods running on different nodes.
+
+* Physical underlay network, composed of L2 switches, L3 routers, load balancers for example, transport packets between nodes. 
```
```suggestion
* An _underlay network_ that the nodes are connected to. For example, if you run secure shell
  (SSH) on a node, the SSH daemon is typically listening on an address associated with this
  underlay network. The kubelet also uses the underlay network to communicate with the
  control plane.
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+
+Figure 2 illustrates an example of localhost communications between two containers on the same pod. You define a pod 1 network namespace that provides pod 1 with its own isolated network stack. Your containers on pod 1 share a single IP and MAC address; Containers on pod 1 are configured with separate port numbers.
+
+{{< figure src="/docs/images/k8net-localhost.drawio.svg" alt="k8s localhost" class="diagram-large" caption="Figure 2. Example of container communications  within the same pod using localhost:port number" >}}
```
```suggestion
{{< figure src="/docs/images/k8net-localhost.drawio.svg" alt="Line drawing diagram showing two containers in a Pod communicating" class="diagram-large" caption="Figure 2. Example of container communications  within the same pod using localhost:port number" >}}
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+
+{{< figure src="/docs/images/k8net-localhost.drawio.svg" alt="k8s localhost" class="diagram-large" caption="Figure 2. Example of container communications  within the same pod using localhost:port number" >}}
+
+Your containers within a pod must coordinate port usage, but this is no different from processes in a VM.  
```
How about:
```suggestion
When you define a Pod, you must make sure that the different containers in that Pod do not conflict
over network port use; this is no different from processes in a VM, or daemons running directly on a physical server machine.
```
?

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+
+How you implement this is a detail of the particular container runtime you employ in your environment.
+
+It is possible to request ports or call host ports on the node itself which forward packets to your pod. This is a very niche operation. How you implement this is a detail of the container runtime. The pod itself is blind to the existence or non-existence of host ports.
```
I would not mention that detail here; I'm not even sure how you configure it within Kubernetes.

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+
+* Communication between the pod network namespaces and root network namespace. The establishment of a veth p2p links accomplishes this.
+
+* L2 or L3 networking between the pods. This example uses a virtual L2 bridge. Your implementation might use a different approach implemented in a network plugin.
```
```suggestion
* Layer 3 networking between the pods. This example uses a virtual layer 2 bridge that connects all
   the Pods on a node. Your cluster might use a different approach.
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+ 
+### Pods on different nodes
+
+Your Kubernetes deployment might place pods on a network of nodes. With this pod topology, you can distribute your workloads across many pods to increase scale and resiliency. Inter-pod networking between distinct nodes becomes a crucial function of your cluster network. 
```
```suggestion
Kubernetes is designed to run at scale, with at least two nodes (for resilience) or even thousands of nodes.
No matter how many nodes you have, Kubernetes is designed to support inter-pod networking between
distinct nodes, without any host being a single point of failure. Networking between pods is  a crucial
function of your cluster.
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+Network plugins support container networking across pods and nodes in your cluster environment. You can think of network plugins as the software responsible for container runtime interface attachment and detachment, pod IP address allocation and network function configuration such as tunnel setup, BGP and IP routing.  
+
+{{< note >}}
+Network plugins are generally referred to as CNI plugins.
```
(nit)
```suggestion
Network plugins are often referred to as CNI plugins.
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+Network plugins are generally referred to as CNI plugins.
+{{< /note >}}
+
+CNI plugins adhere to the specifications defined in the [container network interface](https://www.cni.dev/). The rational behind the development of the CNI was to avoid code duplication, simplify container network configuration and operations, and guarantee cluster compatibility. Most importantly, it offers the Kubernetes community a uni specification for building and deploying container networki ...
```
```suggestion
CNI plugins adhere to the specifications defined in the [container network interface](https://www.cni.dev/).
The rationale behind the development of the CNI was to avoid code duplication, simplify container network
configuration and operations, and guarantee cluster compatibility.
Most importantly, CNI offers the Kubernetes community a unified specification for implementing
container networking functions in cluster networks.
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+
+The CNI provides a [specification](https://www.cni.dev/docs/spec/), a Go [library for CNI application and integration](https://github.com/containernetworking/cni/blob/main/libcni) and a [command-line tool](https://github.com/containernetworking/cni/blob/main/cnitool) for executing CNI plugins. You can find reference plugins and a template in a [separate repository](https://github.com/containernet ...
+
+You will not find a single CNI plugin that addresses all of your requirements. In practice, you will likely employ multiple plugins that in some cases can be "chained" together to support your environment. 
```
```suggestion
CNI plugins are designed to work as part of the overall picture. Some provide basic utility functions - such
as `ptp` to provide a Linux veth pair, or `loopback` to make localhost work. Other network plugins are a
full implementation of pod-to-pod networking, that perhaps rely on the basic utility plugins to work.
You will typically not find a single CNI plugin that addresses all of your requirements. In practice,
you will likely employ multiple plugins that in some cases can be "chained" together to support your environment.
```
?

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+
+* [Reference plugins](https://www.cni.dev/plugins/current/#reference-plugins)
+
+* [3rd party plugins](https://github.com/containernetworking/cni#3rd-party-plugins)
```
```suggestion
* Selection of [3rd party plugins](https://github.com/containernetworking/cni#3rd-party-plugins)
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+
+* [3rd party plugins](https://github.com/containernetworking/cni#3rd-party-plugins)
+
+* [Introduction to CNI](https://youtu.be/YjjrQiJOyME)
```
(nit)
```suggestion
* [Introduction to CNI](https://youtu.be/YjjrQiJOyME) (video) on YouTube
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+Your Kubernetes deployment might place pods on a network of nodes. With this pod topology, you can distribute your workloads across many pods to increase scale and resiliency. Inter-pod networking between distinct nodes becomes a crucial function of your cluster network. 
+
+
+#### Container network interface
```
I would move this section to be either:
- a 2nd level heading, that comes just before `## Design principles for Kubernetes networking`
- a new separate page

In place of the moved text, put a one-paragraph summary and a link to its new home.

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+
+CNI plugins adhere to the specifications defined in the [container network interface](https://www.cni.dev/). The rational behind the development of the CNI was to avoid code duplication, simplify container network configuration and operations, and guarantee cluster compatibility. Most importantly, it offers the Kubernetes community a uni specification for building and deploying container networki ...
+
+The CNI provides a [specification](https://www.cni.dev/docs/spec/), a Go [library for CNI application and integration](https://github.com/containernetworking/cni/blob/main/libcni) and a [command-line tool](https://github.com/containernetworking/cni/blob/main/cnitool) for executing CNI plugins. You can find reference plugins and a template in a [separate repository](https://github.com/containernet ...
```
```suggestion
Along with the CNI [specification](https://www.cni.dev/docs/spec/), there is a Go library,
[libcni](https://github.com/containernetworking/cni/blob/main/libcni) for CNI application and integration,
and a [command-line tool](https://github.com/containernetworking/cni/blob/main/cnitool) for executing
CNI plugins. You can find reference plugins and a template in a [separate repository](https://github.com/containernetworking/plugins) if you wish to build your own.
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+
+#### Pods on different nodes using a virtual overlay network
+
+This example uses a virtual overlay network to support inter-pod networking.  
```
```suggestion
Because the two pods in this scenario are on two different nodes, packets first enter a local
virtual layer 2 bridge, then transit over an overlay network, and finally the layer
2 bridge on the target node. Your cluster might use a different approach.
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+
+This example uses a virtual overlay network to support inter-pod networking.  
+
+{{< figure src="/docs/images/k8net-net-overlay.drawio.svg" alt="k8s pods virtual overlay" class="diagram-large" caption="Figure 4. Example of Pod 1 - Pod 2 networking on different nodes using a virtual overlay network" >}}
```
```suggestion
{{< figure src="/docs/images/k8net-net-overlay.drawio.svg" alt="Diagram showing a virtual overlay network tunnel between two nodes; the tunnel is highlighted in red" class="diagram-large" caption="Figure 4. Example of Pod 1 - Pod 2 networking on different nodes using a virtual overlay network" >}}
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+The virtual overlay approach in this example offers several advantages: 
+
+* Isolates the pod network from the physical underlay network. 
+* Enhanced security with a CNI plugin that handles encrypted communications such as IPsec.
```
```suggestion
* Enhanced security: the tunnel traffic is encrypted, with GENEVE packets protected by IPsec.
  Again, your cluster might use a different design; this is just an example.
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+* Isolates the pod network from the physical underlay network. 
+* Enhanced security with a CNI plugin that handles encrypted communications such as IPsec.
+ 
+A disadvantage to this approach is CNI plugin management, tunnel encap/decap performance overhead and network troubleshooting.
```
```suggestion
A disadvantage to this approach is CNI plugin management, the performance overhead of encapsulation/decapsulation, the management overhead for protecting and distributing encryption
key material, and more complex network troubleshooting.
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+ 
+A disadvantage to this approach is CNI plugin management, tunnel encap/decap performance overhead and network troubleshooting.
+
+For more information on virtual overlay network implementations, see [Flannel](https://github.com/flannel-io/flannel) and [Weavenet](https://www.weave.works/docs/net/latest/overview/). 
```
```suggestion
We avoid linking to specific implementations that are not part of Kubernetes, other than where required
for a cluster to operate. This doesn't feel like we'd have grounds for an exception.
```

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+For more information on virtual overlay network implementations, see [Flannel](https://github.com/flannel-io/flannel) and [Weavenet](https://www.weave.works/docs/net/latest/overview/). 
+
+
+#### Pods on different nodes using a physical underlay network
+
+This example uses the _physical underlay network_ to support inter-pod networking. 
+
+{{< figure src="/docs/images/k8net-net-phys-underlay.drawio.svg" alt="k8s pods physical underay" class="diagram-large" caption="Figure 5. Example of Pod 1 - Pod2 networking on different nodes using a physical underlay network" >}}
+
+There several advantages to the physical underlay approach in this example. 
+
+* No need for you to deploy and manage a separate virtual overlay network.   
+* Leverage existing deployment routing protocols such as BGP, address management and operational troubleshooting.  
+
+A disadvantage you might consider is the additional work to protect the pod network.
+
+For more information on physical network underlay implementations, see [Calico BGP](https://github.com/projectcalico/calico) and [Cilium native-routing](https://docs.cilium.io/en/v1.13/network/concepts/routing/#native-routing). 
```
:warning: This part doesn't look right. The underlay is jargon to describe the network fabric that the overlay network runs on top of.

For example: _imagine that you get 7 Raspberry Pi computers, plug 'em together with Ethernet cables and switches. That's going to be your underlay network. You can touch it and unplug it, if you want to._

_Now set up [Tailscale](https://tailscale.com/), or something like it. On each Pi._

_Get 7 more Raspberry Pis and connect them to a different subnet, again using Ethernet cables etc. Put them on the same Tailscale-managed overlay network. The overlay lets all the nodes talk to each other, even if the two sets of computers are in different rooms / buildings / countries._

Anyway, if you don't set up an overlay VPN, you've not got an underlay. What you have in the no-VPN, all-in-one-apartment case is “a network”.

The heading to use here could be “**#### Pods on different nodes without using an overlay network**”

Before we get to that heading, though, there is a use for the underlay. You need to talk to the Tailscale server. And it's how you set up basic services like NTP for time, or maybe SSH for diagnostics. We might want to hint at that detail before moving on.

In the example with a GENEVE overlay, the underlay network provides:
- transit for the GENEVE-encapsulated overlay traffic
- communication between nodes and control plane
- NTP (we might as well mention it, accurate time sync is a useful thing to have)
  some VM providers let you sync time from the host machine, mind
- IPsec management (IKEv2 and so forth)

and probably also provides:
- a way for load balancers to reach NodePorts on the VMs

I wouldn't mention BGP; it's a complexity best avoided here.

---

Maybe this scenario is best parked for now. How about explaining the scenario where the kubelet makes an HTTP request to a Pod to check whether it is alive? That's a host-to-pod scenario.

There's also a node-to-control-plane scenario you might want to draw; this'll use the underlay network only.

chrismetz09 replied:
The "underlay" jargon is defined/mentioned in multiple documents such as [Geneve: Generic Network Virtualization Encapsulation RFC 8926](https://datatracker.ietf.org/doc/rfc8926/), [An Architecture for Data-Center Network Virtualization over Layer 3 (NVO3)](https://datatracker.ietf.org/doc/html/rfc8014) and many more. 

IMHO, the overlay/underlay paradigm is a simple pattern that new readers can make sense of, and one that is used in K8s clusters today. I think new diagrams might help visualize this better.

sftim replied:
It's fine to have an underlay, but it only makes sense when we're also talking about an overlay. If that was the intent, I think the underlay diagram is fine to keep, but with changes - I think from a technical perspective it's not quite telling the right story.

chrismetz09 replied:
Sounds good. Let me hack some changes to better explain the underlay question and a few others you pointed out.

sftim reviewed (commented):
Thanks.

This is a lot to cover. We couldn't merge this right now; it would mislead people. However, it's a massive step in the right direction.

If you'd like to, you could carve out just the CNI section into a new page that goes into its own smaller PR.

chrismetz09 commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,214 @@ description: >
+Figure 1 illustrates the Kubernetes network architecture and accompanying components used in the examples below. 
+
+{{< figure src="/docs/images/k8net-net-arch2.drawio.svg" alt="k8s net arch" class="diagram-large" caption="Figure 1. K8s Network Architecture and Components" >}}
+The architecture and components consist of the following:
```
Mm ... With my new reader hat on, I might ask "Example of what? Where did those examples come from?" 

All network systems have an architecture. Maybe we call it something else like "K8s Network Model Examples". Diagrams can be simplified/streamlined to keep it simple for the new reader in this first PR with network pictures. Now new readers can dive into the different examples or use-cases.

sftim replied:
I'll try an analogy. If you're trying to draw up the [MVC](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller) pattern, you wouldn't mention the implementation-specific details such as [`NSEditorRegistration`](https://developer.apple.com/documentation/appkit/nseditorregistration). You can do MVC without using Apple's ecosystem.

However, you also might want to provide a non-abstract example to help people learn, and that's fine.

When I commented, there were a bunch of implementation specifics such as veth pairs. I want to avoid implying that these are always part of the architecture.

chrismetz09 commented:
Still working on text and figure cleanup. Want to add other K8s functions that employ physical network independent virtual overlay.

chrismetz09 commented:
https://deploy-preview-39890--kubernetes-io-main-staging.netlify.app/docs/concepts/services-networking/

1. added diagram up front to avoid "text wall greeting". 
2. Trimmed terminology

Still working other sections.

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,196 @@ description: >
+
+Figure 2 illustrates an example of localhost communications between two containers on the same pod, Pod 1, running on a single node, Node 1. 
+
+{{< figure src="/docs/images/k8s-localhost-02.drawio.svg" alt="k8s localhost" class="diagram-large" caption="Figure 2. Container communications within the same pod using localhost:port number" >}}
```
```suggestion
{{< figure src="/docs/images/k8s-localhost-02.drawio.svg" alt="Kubernetes local networking within a Pod" class="diagram-large" caption="Figure 2. Container communications within the same pod using localhost:port number" >}}
```

chrismetz09 replied:
Changed added manually.

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,196 @@ description: >
+
+Figure 3 illustrates an example of two pods talking each other on the same node. 
+
+{{< figure src="/docs/images/k8net-PodSameHost03.drawio.svg" alt="k8s pods samehost" class="diagram-large" caption="Figure 3. Example of Pod 1 and Pod 2 on Node 1 communicating with each other" >}}
```
(nit)
A longer alt attribute would help.

sftim reviewed (commented):
Some more feedback. Do you think this is ready to review & merge @chrismetz09 ?

chrismetz09 commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,196 @@ description: >
+
+Every [pod](/docs/concepts/workloads/pods/) in your cluster gets its own unique cluster-wide IP address called a _pod IP_. 
+
+If you have deployed an [IPv4/IPv6 dual stack](/docs/concepts/service-networking/dual-stack.md) cluster, then you must allocate pod IPs for IPv4 and IPv6.  a can allocate pod IP) ;This is performed on per [_address family_](https://www.iana.org/assignments/address-family-numbers/address-family-numbers.xhtml), so a cluster that uses IPv4 and IPv6 networking assigns one IPv4 address and IPv6 addres ...
```
Definitely not ready. Couple of more days needed.

chrismetz09 commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,196 @@ description: >
+without network address translation (NAT).
+* Pods can communicate with other pods on the same or separate nodes using layer 2 bridging, native layer 3 (IPv4, IPv6) networking, tunneling, or other techniques for conveying packets from one pod to another.
+
+Some platforms, such as Linux, support pods running in the host network. Pods attached to the host network of a node can still communicate with all pods on all nodes without NAT.
```
Added windows. Change host network to `hostNetwork`?

chrismetz09 commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,196 @@ description: >
+
+This model offers the following advantages:
+
+* Clean, backwards-compatible approach where `Pods` can be treated 
```
Change added manually. Should I capitalize Pods for all instances in this PR?

sftim replied:
> capitalize Pods for all instances in this PR?

Only when you're talking about Pods as things in the API. For example, if you're talking about a set of cgroups and namespaces that hold some containers, then no: call it a “pod”, lowercase.

chrismetz09 replied:
got it!

nate-double-u commented:
I've asked for a [technical review over in #sig-network](https://kubernetes.slack.com/archives/C09QYUH5W/p1683157062116469).

As for the deploy, it looks like it [timed out without much of an error](https://app.netlify.com/sites/kubernetes-io-main-staging/deploys/6452da66262e0d0008e2e48f). To re-trigger the checks for this, I think you can you either `git commit --amend --date=now` or `git commit --allow-empty` and force push to your fork

aojea commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,195 @@ description: >
+* Containers in same Pod can communicate with each other.
+* Pods can communicate with all other Pods on the same or separate [nodes](/docs/concepts/architecture/nodes/)
+without network address translation (NAT).
+* Pods can communicate with other Pods on the same or separate nodes using layer 2 bridging, native layer 3 (IPv4, IPv6) networking, MPLS, segment routing, tunneling, or other techniques for conveying packets from one Pod to another.
```
I prefer the existing paragraph, why do we need to enumerate all the network technologies?

chrismetz09 replied:
Some new readers from the networking world or operators who have deployed SRv6, for example, may wish to know if a protocol can accommodate K8s network model. Vendors, operators, students, network architects, etc. have discussed these topics.

chrismetz09 replied:
Also Kubernetes added IPv6 after IPv4. See your point though.

aojea commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,195 @@ description: >
 
-   * pods can communicate with all other pods on any other [node](/docs/concepts/architecture/nodes/) 
-     without NAT
-   * agents on a node (e.g. system daemons, kubelet) can communicate with all
```
this is important, is reflected in other place?

chrismetz09 replied:
Yes but I'll put the "agents on a node" bullet back in.

aojea commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,195 @@ description: >
+
+Inter-container connectivity on the same Pod makes use of pod IPs, network namespaces and localhost networking.
+
+Pods run in their own [network namespace](https://man7.org/linux/man-pages/man7/network_namespaces.7.html). All the containers in a Pod share this network namespace. (Network namespaces are not the same as the Kubernetes {{< glossary_tooltip text="namespace" term_id="namespace" >}} concept).
```
yeah, and there is windows too :/

aojea commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,195 @@ description: >
+
+Pods run in their own [network namespace](https://man7.org/linux/man-pages/man7/network_namespaces.7.html). All the containers in a Pod share this network namespace. (Network namespaces are not the same as the Kubernetes {{< glossary_tooltip text="namespace" term_id="namespace" >}} concept).
+
+Figure 2 illustrates an example of localhost communications between two containers on the same Pod. 
```
is not very illustrative, containers will use sockets directly to connect to the other socket, I really can't think about good way. of illustrate it

chrismetz09 replied:
Researching. I have only read about localhost communications.

aojea commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,195 @@ description: >
+
+### Pods on the same node
+
+Inter-pod connectivity on the same node utilize pod IPs, Pod and root network namespaces, [veth links](https://man7.org/linux/man-pages/man4/veth.4.html) and a [virtual L2 bridge](https://en.wikipedia.org/wiki/Data_link_layer).
```
this is an implementation details, not all of them use virtual bridges, actually that should be discourage on favor of ptp interfaces for security ...

chrismetz09 replied:
I had this: `L2 or L3 networking between the Pods. This example uses a virtual L2 bridge. You might use a different approach supported by your network plugin.`

I'll incorporate your points.

aojea commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,195 @@ description: >
+
+Inter-pod connectivity on the same node utilize pod IPs, Pod and root network namespaces, [veth links](https://man7.org/linux/man-pages/man4/veth.4.html) and a [virtual L2 bridge](https://en.wikipedia.org/wiki/Data_link_layer).
+
+Pods operate in their own network namespace and have their own IP and MAC addresses. You can think of this as multiple distinct IP computers (pods), that wish to talk with each other through a L2 bridge. Inter-pod networking on the same node is no different.
```
the bridge is an implementation details and is not the best one

chrismetz09 replied:
Seems like  it sufficed as a basic and well-understood implementation example. I'll change text to clarify your points.

sftim replied:
We can teach that a bridge is a common / simple implementation; we must avoid implying that it's mandatory. Even _Ethernet_ itself isn't part of the Kubernetes network model - virtual or otherwise.

aojea commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,195 @@ description: >
+
+In all scenarios involving inter-node Pod networking (physical or virtual), you will have a physical network in place. This provides you with communications between different network resources needed for K8s and non-K8s application support, APIs, metrics collection, operations and troubleshooting. 
+
+For a K8s-specific example, you could have a process call the control plane API server to issue an HTTP request to a node kubelet. 
```
I can not fully understand this sentence sorry?

aojea replied:
the way this can work is the kubelet register itsefl with the apiserver creating the corresponding node object that represent itself.
It can also happen that the apiserver connects to the kubelet to perform some operations like "kubectl exec" per example

sftim replied:
You might only be configuring a virtual network; for example, in the cloud, where the underlying physical network is abstracted away.

aojea commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,195 @@ description: >
+
+{{< figure src="/docs/images/k8s-api-to-kubelet-example.drawio.svg" alt="diagram of k8s API server to node kubelet" class="diagram-large" caption="Figure 4. Example of control plane to node communications using the physical network" >}}
+
+In addition to addressing your K8s inter-node Pod networking requirements, your physical network can support NTP, IPsec IKEv2, OAM tasks and others specific to your environment.
```
I don't know if we should go into those details, is just a network, people can run more things, maybe you want to highlight that is not necessary a kubernetes only network?

chrismetz09 replied:
Got it.

aojea commented:
I think that this is better fit for a blog

chrismetz09 commented:
Working to resolve build/deploy preview problem per #35780 . Once fixed, should make reviewing easier.

Tx @aojea, for the initial review. Still working it.

sftim commented:
If you rebase against _main_ that might help with the previews @chrismetz09

chrismetz09 commented:
@sftim, just want to confirm correct commands to rebase against main. 

Starting from my dev branched called `net-model-diagrams`, I type in:

`git fetch && git checkout main && git pull origin main`
`git checkout <name of your branch> && git rebase -i main`

chrismetz09 commented:
Preview available at https://deploy-preview-39890--kubernetes-io-main-staging.netlify.app/docs/concepts/services-networking/. This will help reviewers.

Currently working on new commits suggested by @aojea.

sftim commented:
For https://deploy-preview-39890--kubernetes-io-main-staging.netlify.app/docs/concepts/services-networking/#containers-on-the-same-pod, can we omit mentioning a MAC address? MAC addresses are a (commonly seen) implementation detail.

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,195 @@ description: >
+
+Pods operate in their own network namespace and have their own IP and MAC addresses. You can think of this as multiple distinct IP computers (pods), that wish to talk with each other through a L2 bridge. Inter-pod networking on the same node is no different.
+
+This example of inter-pod networking on the same node requires:
+
+* Root network namespace for the node. 
+
+* Veth links enabling communication between the Pod network namespace and root network namespace. 
```
Whilst this is true, I'm worried that readers will infer that _all_ inter-pod packet networking requires a veth.

chrismetz09 replied:
There are a large number of K8s network blogs already out there. Would we add any value? New reader sees the components, wonders how this could work with simple examples and then off they go to google or K8s blog? 

I have always appreciated "one stop shop" for architecture and then implementation examples. Now I have "jumping off point" to dig deeper. Networking is hard; examples in close proximity is a must for learning, IMO.  

We have [K8s Components](https://kubernetes.io/docs/concepts/overview/components/). Perhaps I could zoom in to the nodes and add the core network components?

sftim replied:
I like this diagram; it's just that I'm confident lots of people will misunderstand what we're saying, unless we reframe the explanation.

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,195 @@ description: >
+
+#### Pods on different nodes using a physical network
+
+This example uses Pod and root network namespaces, veth links, network plugins and the physical network. 
+
+Figure 5 illustrates an example where Pods on different nodes use the physical network to communicate.
+
+{{< figure src="/docs/images/k8s-net-phys-net.drawio.svg" alt="diagram of inter-node Pod networking using physical network " class="diagram-large" caption="Figure 5. Example of Pod 1 - Pod 2 networking on different nodes using a physical network" >}}
+
+The network plugin exposes Pods (pod IP) to the physical network addressing and forwarding. In essence your Pods become IP host end-points reachable by all other end-points on the physical network. You can leverage existing network transport, address management, performance, security and operational best-practices to support your Pod networking requirements.
+
+To ensure Pod access integrity, you should apply [pod security](/docs/concepts/security/pod-security-standards) measures along with one or more network functions designed to protect hosts from malicious attacks. Examples include packet filters, address management, network segmentation, authentication and encryption. 
+
+The physical network approach offers you several advantages:
+
+* Common physical network to handle K8s and non-K8s traffic.
+* Optimal network performance.
+* Enhanced security with a CNI plugin that handles encrypted communications such as IPsec or TLS.
+
+The challenges to you include CNI plugin management and non-K8s network operations impact to your cluster.
+
+For more information on two examples of physical network implementations, see [Calico BGP](https://github.com/projectcalico/calico) and [Cilium native-routing](https://docs.cilium.io/en/v1.13/network/concepts/routing/#native-routing).
```
There's a detail missing here: the bridging. The node is using a software bridge to make each Pod appear as if it has a physical connection. The node still only has one network interface.

chrismetz09 replied:
Is the software bridge the virtual L2 bridge or something different?

sftim replied:
The text is what could mislead; eg: _The network plugin exposes Pods (pod IP) to the physical network addressing and forwarding._

That's where I'd mention the use of a bridge. There actually are designs where each Pod gets its own network interface that can talk directly to things outside the node. This isn't one of those designs.

sftim commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,195 @@ description: >
+
+#### Pods on different nodes using a virtual overlay network
+
+You can address your inter-node pod networking requirements by deploying a [virtual overlay network](https://book.systemsapproach.org/applications/overlays.html). This approach employs virtual tunnels to form a separate logical _overlay network_ for the express purpose of transporting packets from one pod to another.
+
+Figure 6 illustrates the notion of a virtual overlay network running through (or on top of) a physical underlay network.   
+
+{{< figure src="/docs/images/k8s-net-vert-overlay-big2.drawio.svg" alt="diagram of virtual overlay network running on top of a physical underlay network." class="diagram-large" caption="Figure 6. Example of a virtual overlay network" >}}
+ 
+Pods 1, 2 and 3 are networked together through a mesh of virtual tunnels. Your encapsulated pod packets injected into a virtual tunnel are transported from one node to another across your physical underlay network. 
+
+{{< note >}}
+A physical network is sometimes referred to as a _physical underlay network_. This because it is transporting virtual overlay network packets.   
+{{< /note >}}
+
+Figure 7 shows an example of a virtual overlay network composed of two pods running on separate nodes and connected via virtual overlay network tunnel. It uses pod and root network namespaces, network plugins, tunnel encapsuation/decapsulation, and a physical underlay network.
+ 
+{{< figure src="/docs/images/k8s-net-virtual-overlay.drawio.svg" alt="diagram of 2 x k8s pods on separate nodes using virtual tunnels" class="diagram-large" caption="Figure 7. Example of Pod 1 - Pod 2 networking on different nodes using a virtual overlay network tunnel" >}}
+
+In this example, the network plugin has set up a _virtual overlay network_. Pod packets sourced on one node and destined for Pods 
+on a different node are encapsulated and sent over a tunnel mechanism such as [GENEVE](https://en.wikipedia.org/wiki/Generic_Network_Virtualization_Encapsulation). 
+
+The destination node retrieves the encapsulated pod packet, strips off the encap header, and then sends it to the target Pod, providing that there are no
+[NetworkPolicy](/docs/concepts/services-networking/network-policies/) restrictions to block that packet.
```
For this example, consider sending the VPN traffic over a second host interface. I'd like to make the point that nodes can have more than one interface.

chrismetz09 replied:
Got it!

sftim reviewed (commented):
IMO:

- the different examples of Pod networking implementations would work well in a blog article
- explaining the **core elements** of Kubernetes networking is a good thing for https://k8s.io/docs/concepts/services-networking/

chrismetz09 commented:
Some notes for working an architecture diagram. 

1. Modify implementation examples to reflect architecture. So for example veth or ptp becomes just generic pt-pt link. Or namespace pt-pt link? Virtual L2 bridge becomes Local Pod Forwarder?
2. More emphasis on IP-per-pod model.
3. Network plugins can be CNI plugins?
4. Node can have more than one interface.

I'll post draft architecture diagram for further discussion.

sftim commented:
We don't mandate that all network plugins are CNI plugins, but there aren't any counterexamples in the wild that work with pure upstream Kubernetes.

chrismetz09 commented:
Partial hack net mod arch: build off of [K8s Components](https://kubernetes.io/docs/concepts/overview/components/). . Components diagram does not include containers, pods or networking. 

IMHO, we should start simple and subsequent PRs (lotsa them) can zoom in and add new pieces. I feel complexity creep seeping in for this first cut.  
 
![k8s-net-model-arch drawio](https://github.com/kubernetes/website/assets/5543858/9006f33b-b45d-451a-a405-54799a988dde)

chrismetz09 commented:
Excellent. I'll make the updates, tx.

chrismetz09 commented:
![k8s-network-all-template-k8s-net-model-arch2 drawio](https://github.com/kubernetes/website/assets/5543858/1220d331-b6b7-4765-9bc8-e4a1f4f2db8f)

Revised figure per @sftim suggestions. So what's the plan on this PR journey? Current website doc offers minimal k8s network coverage (just text). IMHO, interested K8s readers need more. 

My suggestions:

1. Explain K8s Network Components using new figure just as K8s components are explained [here](https://kubernetes.io/docs/concepts/overview/components/). This gives us a K8s network architecture. Updates can come in future PRs. 

2. Then we have IMPLEMENTATION EXAMPLES for container-to-container same pod, pod same host and pod different hosts using tunnels. We will be 100% clear that these are EXAMPLES so as not to confuse the reader.

This would be a good start and then we can iterate with future PRs.

Thoughts?

sftim reviewed (commented):
I broadly support the approach in https://github.com/kubernetes/website/pull/39890#issuecomment-1568728147

Can we save the implementation examples for a follow-up PR? And initially leave them out entirely.

Please do save the work done here, it's useful but just not quite right for a merge as it stands.

Also, please write a description for this PR. That helps us to review the outcome against the described intent.

chrismetz09 commented on content/en/docs/concepts/services-networking/_index.md:
```diff
@@ -7,44 +7,195 @@ description: >
+
+Inter-container connectivity on the same Pod makes use of pod IPs, network namespaces and localhost networking.
+
+Pods run in their own [network namespace](https://man7.org/linux/man-pages/man7/network_namespaces.7.html). All the containers in a Pod share this network namespace. (Network namespaces are not the same as the Kubernetes {{< glossary_tooltip text="namespace" term_id="namespace" >}} concept).
```
Okay. I have not pushed a commit in several weeks.

sftim replied:
Sorry, might not have been clear - by “previous version” I mean the red colored text from that latest actual commit (vs. my suggestion).

They are both OK to merge, even if I prefer making the Linux detail clear.

chrismetz09 commented:
[Final Preview](https://deploy-preview-39890--kubernetes-io-main-staging.netlify.app/docs/concepts/)

- Compressed to and cleaned up text for model components (diagram), requirements, and diagrams for implementation examples. Everything else commented out in content file.
- added URL for diagrams at diagrams.net.
- added PR description. 
- closed PR.

chrismetz09 commented:
Work available for future PR consideration.
